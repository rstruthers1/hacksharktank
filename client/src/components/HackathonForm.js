import React from 'react';
import {Controller, useForm} from 'react-hook-form';
import {yupResolver} from '@hookform/resolvers/yup';
import * as Yup from 'yup';
import DatePicker from 'react-datepicker'
import './CenteredForms.css';
import 'react-datepicker/dist/react-datepicker.css'
import {getTodayAtMidnight} from "../utils/dateTimeUtils";


const HackathonForm = () => {

    const validationSchema = Yup.object().shape({
        eventName: Yup.string().required('Event Name is required').min(3, 'Event Name must be at least 3 characters long'),
        startDate: Yup.date().required('Start Date is required').typeError('Start Date is required').min(getTodayAtMidnight(), 'Start Date must be today or later'),
        endDate: Yup.date().required('End Date is required').typeError('End Date is required').min(Yup.ref('startDate'), 'End Date must be after Start Date')
    });

    // Initialize react-hook-form methods
    const {control, register, handleSubmit, watch, formState: {errors}} = useForm({
        resolver: yupResolver(validationSchema),
        mode: "onChange"
    });

    const currentStartDate = watch('startDate', false);

    const onSubmit = async data => {
        // Handle your form submission here
        console.log(data);
    }

    return (
        <div className="centeredForm">
            <form onSubmit={handleSubmit(onSubmit)}>
                <h2>Create Hackathon Event</h2>
                <div className="formField">
                    <label htmlFor="eventName">Event Name</label>
                    <input type="text" {...register('eventName')}/>
                    <p className="error">{errors.eventName?.message}</p>
                </div>
                <div className="formField">
                    <label htmlFor="startDate">Start Date</label>
                    <Controller
                        control={control}
                        name='startDate'
                        render={({ field }) => (
                            <DatePicker
                                placeholderText='MM/dd/yyyy'
                                onChange={(date) => {field.onChange(date)}}
                                selected={field.value}
                                dateFormat='MM/dd/yyyy'
                                minDate={getTodayAtMidnight()}
                            />
                        )}
                    />
                    <p className="error">{errors.startDate?.message}</p>
                </div>
                <div className="formField">
                    <label htmlFor="endDate">End Date</label>
                    <Controller
                        control={control}
                        name='endDate'
                        render={({ field }) => (
                            <DatePicker
                                placeholderText='MM/dd/yyyy'
                                onChange={(date) => field.onChange(date)}
                                selected={field.value}
                                dateFormat='MM/dd/yyyy'
                                minDate={currentStartDate ? currentStartDate : getTodayAtMidnight()}
                            />
                        )}
                    />
                    <p className="error">{errors.endDate?.message}</p>
                </div>
                <button type="submit" className="submit-btn">Create Event</button>
            </form>
        </div>
    );
};

export default HackathonForm;
