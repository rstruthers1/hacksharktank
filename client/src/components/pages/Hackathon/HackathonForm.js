import {Controller, useForm} from "react-hook-form";
import DatePicker from "react-datepicker";
import {getMinEndDate, getTodayAtMidnight, safeStringToDate} from "../../../utils/dateTimeUtils";
import React from "react";
import {yupResolver} from "@hookform/resolvers/yup";
import * as Yup from "yup";

const HackathonForm = ({onSubmit, submitButtonLabel, hackathon}) => {
    const validationSchema = Yup.object().shape({
        eventName: Yup.string().required('Event Name is required').min(3, 'Event Name must be at least 3 characters long').max(256, 'Event Name must be at most 256 characters long'),
        description: Yup.string().required('Description is required').min(3, 'Description must be at least 3 characters long').max(512, 'Description must be at most 512 characters long'),
        startDate: Yup.date().required('Start Date is required').typeError('Start Date is required').min(getTodayAtMidnight(), 'Start Date must be today or later'),
        endDate: Yup.date().required('End Date is required').typeError('End Date is required').when('startDate', (startDate, schema) => {
            if (startDate && Array.isArray(startDate) && startDate.length > 0) {
                try {
                    const startDateAsDate = new Date(startDate[0]);
                    const dayAfter = new Date(startDateAsDate.getTime() + 86400000);
                    return schema.min(dayAfter, 'End Date must be at least one day after Start Date');
                } catch (err) {
                    console.error(err);
                    return schema;
                }
            }
            return schema;
        })
    });

    // Initialize react-hook-form methods
    const {control, register, handleSubmit, watch, formState: {errors}} = useForm({
        resolver: yupResolver(validationSchema),
        mode: "onChange",
        defaultValues: {
            eventName: hackathon?.eventName ? hackathon.eventName : '',
            description: hackathon?.description ? hackathon.description : '',
            startDate: safeStringToDate(hackathon?.startDate),
            endDate: safeStringToDate(hackathon?.endDate)
        }
    });

    const currentStartDate = watch('startDate');
    const currentDescription = watch('description');

    return (
        <form onSubmit={handleSubmit(onSubmit)}>
            <div className="formField">
                <label htmlFor="eventName">Event Name <span className="required">*</span></label>
                <input type="text" {...register('eventName')}/>
                {errors.eventName?.message && <p className="error">{errors.eventName?.message}</p>}
            </div>
            <div className="formField">
                <label htmlFor="description">Event Description <span className="required">*</span></label>
                <textarea {...register('description')} />
                <div className="note">{currentDescription ? currentDescription.length : 0}/512 characters</div>
                {errors.description?.message && <p className="error">{errors.description?.message}</p>}
            </div>
            <div className="formField">
                <label htmlFor="startDate">Start Date <span className="required">*</span></label>
                <Controller
                    control={control}
                    name='startDate'
                    render={({field}) => (
                        <DatePicker
                            placeholderText='mm/dd/yyyy'
                            onChange={(date) => {
                                field.onChange(date)
                            }}
                            selected={field.value}
                            dateFormat='MM/dd/yyyy'
                            minDate={getTodayAtMidnight()}
                        />
                    )}
                />
                {errors.startDate?.message && <p className="error">{errors.startDate?.message}</p>}
            </div>
            <div className="formField">
                <label htmlFor="endDate">End Date <span className="required">*</span></label>
                <Controller
                    control={control}
                    name='endDate'
                    render={({field}) => (
                        <DatePicker
                            placeholderText='mm/dd/yyyy'
                            onChange={(date) => field.onChange(date)}
                            selected={field.value}
                            dateFormat='MM/dd/yyyy'
                            minDate={getMinEndDate(currentStartDate)}
                        />
                    )}
                />
                {errors.endDate?.message && <p className="error">{errors.endDate?.message}</p>}
            </div>
            <p><span className="required">*</span> Indicates a required field</p>
            <button type="submit" className="submit-btn">{submitButtonLabel}</button>
        </form>
    )
}

export default HackathonForm;