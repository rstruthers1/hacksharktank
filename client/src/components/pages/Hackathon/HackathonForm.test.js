import {render, screen} from '@testing-library/react';
import renderer from 'react-test-renderer';
import '@testing-library/jest-dom';
import HackathonForm from "./HackathonForm";

describe('HackathonForm', () => {
    it('should render correctly', () => {
        const tree = renderer.create(<HackathonForm/>).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('should render with correct values of the hackathon and submit button label', () => {
        const hackathon = {
            eventName: 'Test Hackathon',
            description: 'This is a test',
            startDate: '2023-11-20T06:00:00.000Z',
            endDate: '2023-11-22T06:00:00.000Z'
        }
        const {container} = render(<HackathonForm onSubmit={jest.fn()} submitButtonLabel="Update" hackathon={hackathon}/>);
        expect(screen.getByText('Update')).toBeInTheDocument();
        const eventNameElement = container.querySelector('input', {value: 'Test Hackathon'});
        expect(eventNameElement).toBeInTheDocument();
        const descriptionElement = container.querySelector('textarea', {value: 'This is a test'});
        expect(descriptionElement).toBeInTheDocument();
        const startDateElement = screen.getByTestId('startDate');
        expect(startDateElement).toBeInTheDocument();
        expect(startDateElement).toHaveValue('11/20/2023');
        const endDateElement = screen.getByTestId('endDate');
        expect (endDateElement).toBeInTheDocument();
        expect(endDateElement).toHaveValue('11/22/2023');
    });

});