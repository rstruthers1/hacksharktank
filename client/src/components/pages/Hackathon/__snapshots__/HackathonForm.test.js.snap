// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HackathonForm should render correctly 1`] = `
<form
  onSubmit={[Function]}
>
  <div
    className="formField"
  >
    <label
      htmlFor="eventName"
    >
      Event Name 
      <span
        className="required"
      >
        *
      </span>
    </label>
    <input
      name="eventName"
      onBlur={[Function]}
      onChange={[Function]}
      type="text"
    />
  </div>
  <div
    className="formField"
  >
    <label
      htmlFor="description"
    >
      Event Description 
      <span
        className="required"
      >
        *
      </span>
    </label>
    <textarea
      name="description"
      onBlur={[Function]}
      onChange={[Function]}
    />
    <div
      className="note"
    >
      0
      /512 characters
    </div>
  </div>
  <div
    className="formField"
  >
    <label
      htmlFor="startDate"
    >
      Start Date 
      <span
        className="required"
      >
        *
      </span>
    </label>
    <div
      className="react-datepicker-wrapper"
    >
      <div
        className="react-datepicker__input-container"
      >
        <input
          className=""
          disabled={false}
          onBlur={[Function]}
          onChange={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          placeholder="mm/dd/yyyy"
          readOnly={false}
          type="text"
          value=""
        />
      </div>
    </div>
  </div>
  <div
    className="formField"
  >
    <label
      htmlFor="endDate"
    >
      End Date 
      <span
        className="required"
      >
        *
      </span>
    </label>
    <div
      className="react-datepicker-wrapper"
    >
      <div
        className="react-datepicker__input-container"
      >
        <input
          className=""
          disabled={false}
          onBlur={[Function]}
          onChange={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          placeholder="mm/dd/yyyy"
          readOnly={false}
          type="text"
          value=""
        />
      </div>
    </div>
  </div>
  <p>
    <span
      className="required"
    >
      *
    </span>
     Indicates a required field
  </p>
  <button
    className="submit-btn"
    type="submit"
  />
</form>
`;
